File: src/2025-02-24-deepclaude-rust/go/exercises/practice/simple-linked-list/simple_linked_list.go
package linkedlist

import "errors"

// Define the List and Element types here.
type Element struct {
    Value int
    Next  *Element
}

type List struct {
    head *Element
    size int
}

func New(elements []int) *List {
	list := &List{}
	for i := len(elements) - 1; i >= 0; i-- {
		list.Push(elements[i])
	}
	return list
}

func (l *List) Size() int {
	return l.size
}

func (l *List) Push(element int) {
	l.head = &Element{Value: element, Next: l.head}
	l.size++
}

func (l *List) Pop() (int, error) {
	if l.head == nil {
		return 0, errors.New("empty list")
	}
	val := l.head.Value
	l.head = l.head.Next
	l.size--
	return val, nil
}

func (l *List) Array() []int {
	arr := make([]int, 0, l.size)
	current := l.head
	for current != nil {
		arr = append(arr, current.Value)
		current = current.Next
	}
	return arr
}

func (l *List) Reverse() *List {
	elements := l.Array()
	for i, j := 0, len(elements)-1; i < j; i, j = i+1, j-1 {
		elements[i], elements[j] = elements[j], elements[i]
	}
	return New(elements)
}

File: src/2025-02-24-deepclaude-rust/java/exercises/practice/simple-linked-list/src/main/java/SimpleLinkedList.java
class SimpleLinkedList<T> {
    private static class Node<T> {
        private final T data;
        private Node<T> next;

        Node(T data) {
            this.data = data;
            this.next = null;
        }

        Node(T data, Node<T> next) {
            this.data = data;
            this.next = next;
        }
    }

    private Node<T> head;
    private int size;

    SimpleLinkedList() {
        head = null;
        size = 0;
    }

    SimpleLinkedList(T[] values) {
        this();
        if (values != null) {
            for (T value : values) {
                Node<T> newNode = new Node<>(value);
                if (head == null) {
                    head = newNode;
                } else {
                    Node<T> current = head;
                    while (current.next != null) {
                        current = current.next;
                    }
                    current.next = newNode;
                }
                size++;
            }
        }
    }

    void push(T value) {
        Node<T> newNode = new Node<>(value, head);
        head = newNode;
        size++;
    }

    T pop() {
        if (head == null) {
            throw new java.util.NoSuchElementException("List is empty");
        }
        T data = head.data;
        head = head.next;
        size--;
        return data;
    }

    void reverse() {
        Node<T> prev = null;
        Node<T> current = head;
        Node<T> next = null;

        while (current != null) {
            next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }
        head = prev;
    }

    @SuppressWarnings("unchecked")
    T[] asArray(Class<T> clazz) {
        Object[] array = new Object[size];
        Node<T> current = head;
        int i = 0;
        while (current != null) {
            array[i++] = current.data;
            current = current.next;
        }
        return (T[]) Arrays.copyOf(array, size, clazz);
    }

    int size() {
        return size;
    }
}
